balancer/pickfirst/pickfirstleaf/pickfirstleaf_ext_test.go

```diff
@@ -851,6 +850,253 @@ func (s) TestPickFirstLeaf_EmptyAddressList(t *testing.T) {
 	}
 }
 
+// TestPickFirstLeaf_HappyEyeballs_TFAfterEndOfList verifies that pickfirst
+// correctly detects the end of the first happy eyeballs pass when the timer
+// causes pickfirst to reach the end of the address list and failures are
+// reported out of order.
+func (s) TestPickFirstLeaf_HappyEyeballs_TFAfterEndOfList(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	timerCh := make(chan struct{})
+	originalTimer := pfinternal.TimeAfterFunc
+	pfinternal.TimeAfterFunc = func(_ time.Duration, f func()) *time.Timer {
+		// Set a really long expiration to prevent it from triggering
+		// automatically.
+		timer := time.AfterFunc(time.Hour, f)
+		go func() {
+			select {
+			case <-ctx.Done():
+			case <-timerCh:
+			}
+			timer.Reset(0)
+		}()
+		return timer
+	}
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	defer func() {
+		pfinternal.TimeAfterFunc = originalTimer
+	}()
+
+	dialer := testutils.NewBlockingDialer()
+	cc, rb, bm := setupPickFirstLeaf(t, 3,
+		grpc.WithDefaultServiceConfig(fmt.Sprintf(`{"loadBalancingConfig": [{"%s":{}}]}`, pickfirstleaf.Name)),
+		grpc.WithContextDialer(dialer.DialContext),
+	)
+	addrs := bm.resolverAddrs()
+	holds := bm.holds(dialer)
+	rb.UpdateState(resolver.State{Addresses: addrs})
+	cc.Connect()
+
+	testutils.AwaitState(ctx, t, cc, connectivity.Connecting)
+
+	// Verify that only the first server is contacted.
+	if got, want := holds[0].Wait(ctx), true; got != want {
+		t.Fatalf("hold[%q].Wait() = %t, want %t", addrs[0], got, want)
+	}
+
+	// Ensure no other servers are contacted.
+	if got, want := holds[1].IsStarted(), false; got != want {
+		t.Fatalf("holds[%q].IsStarted() = %t, want %t", addrs[1], got, want)
+	}
+
+	if got, want := holds[2].IsStarted(), false; got != want {
+		t.Fatalf("holds[%q].IsStarted() = %t, want %t", addrs[2], got, want)
+	}
+
+	// Make the happy eyeballs timer fire twice so that pickfirst reaches the
+	// last address in the list.
+	timerCh <- struct{}{}
+
+	// Verify that the second server is contacted and 3rd isn't.
+	if got, want := holds[1].Wait(ctx), true; got != want {
+		t.Fatalf("hold[%q].Wait() = %t, want %t", addrs[1], got, want)
```

Comment:
I understand the style guide recommendation to identify the function in the test error message, but this is a function on a test utility type and not the function under test. So, maybe something like:
```
	t.Fatalf("Timeout when waiting for second server with address %q to be contacted", addrs[1])
```

Similarly for other error strings where we are calling `hold.Wait` or `hold.IsStarted`.

Reply:
Changed. Using the server numbers instead of the words "first, second etc." as its easier to write copy/paste and write vim macros to make all the changes in one go.